src/main/java/io/cdap/plugin/aggregator/Pivot.java

```diff
@@ -0,0 +1,409 @@
+import com.google.common.collect.Sets;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.cdap.etl.api.InvalidEntry;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.StageConfigurer;
+import io.cdap.cdap.etl.api.batch.BatchAggregator;
+import io.cdap.cdap.etl.api.batch.BatchAggregatorContext;
+import io.cdap.cdap.etl.api.batch.BatchReducibleAggregator;
+import io.cdap.cdap.etl.api.batch.BatchRuntimeContext;
+import io.cdap.cdap.etl.api.lineage.field.FieldOperation;
+import io.cdap.cdap.etl.api.lineage.field.FieldTransformOperation;
+import io.cdap.plugin.aggregator.function.AggregateFunction;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Deduplicate aggregator.
+ */
+@Plugin(type = BatchAggregator.PLUGIN_TYPE)
+@Name("Pivot")
+@Description("Transpose specific rows as columns, and generate aggregate results in those columns")
+public class Pivot extends BatchReducibleAggregator<StructuredRecord, StructuredRecord, AggregateResult,
+  StructuredRecord> {
+  private final PivotConfig config;
+
+  @Nullable
+  private final Integer numPartitions;
+
+  private Schema outputSchema;
+  private List<PivotConfig.FunctionInfo> functionInfos;
+  private Map<String, Set<String>> pivotColumnsAndData;
+  //keep ordered columns name, to generate always same key from columns data models
+  private ArrayList<String> orderedColumnsName;
+  private List<String> columnsProduct;
+
+  private Map<String, String> fieldAliases;
+  private FailureCollector failureCollector;
+
+
+  public Pivot(PivotConfig config, @Nullable Integer numPartitions) {
+    this.config = config;
+    this.numPartitions = numPartitions;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    failureCollector = stageConfigurer.getFailureCollector();
+
+    Schema inputSchema = stageConfigurer.getInputSchema();
+
+    config.validate(failureCollector);
+    if (inputSchema == null || config.containsMacro()) {
+      stageConfigurer.setOutputSchema(null);
+      return;
+    }
+
+    init(failureCollector);
+    validate(inputSchema, failureCollector);
+    outputSchema = generateOutputSchema(inputSchema, config.getPivotRow(), functionInfos, columnsProduct);
+    stageConfigurer.setOutputSchema(outputSchema);
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
```

Comment:
Prepare run should call validation as well. See that in other plugins like google-cloud